src/Storages/StorageMerge.cpp

```diff
@@ -783,12 +849,119 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
 
         /// Subordinary tables could have different but convertible types, like numeric types of different width.
         /// We must return streams with structure equals to structure of Merge table.
-        convertingSourceStream(header, storage_snapshot->metadata, aliases, modified_context, *builder, processed_stage);
+        /// Besides this we add FilterTransform if it is needed to follow row level policies.
+
+        convertingSourceStream(header,
+            storage_snapshot->metadata,
+            aliases,
+            modified_context,
+            *builder,
+            processed_stage,
+            row_policy_data);
     }
 
     return builder;
 }
 
+void ReadFromMerge::RowPolicyData::init(RowPolicyFilterPtr row_policy_filter_ptr_,
+    std::shared_ptr<DB::IStorage> storage,
+    ContextPtr local_context)
+{
+    if (row_policy_filter_ptr_)
+    {
+        row_policy_filter_ptr = row_policy_filter_ptr_;
+
+        ASTPtr expr = row_policy_filter_ptr->expression;
+
+        auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+        auto storage_columns = storage_metadata_snapshot->getColumns();
+        auto needed_columns = storage_columns.getAllPhysical();
+
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
+        auto expression_analyzer = ExpressionAnalyzer{expr, syntax_result, local_context};
+
+        actions_dag = expression_analyzer.getActionsDAG(false /* add_aliases */, false /* project_result */);
+        filter_actions = std::make_shared<ExpressionActions>(actions_dag,
+            ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+        filter_column_name = namesDifference(filter_actions->getSampleBlock().getNames(), filter_actions->getRequiredColumns());
+    }
+}
+
+// Add columns that needed to evaluate row policies
+// SELECT x from t  if  t has row policy
+void ReadFromMerge::RowPolicyData::extendNames(Names & names)
+{
+    assert(row_policy_filter_ptr);
+    ASTPtr expr = row_policy_filter_ptr->expression;
+
```

Comment:
It's not necessary to use binary search here. The number of columns is not that big anyway.
Just store the input columns in the `init()` function and then use an ordinary linear search:

```
    for (const auto & filter_input_column : filter_input_columns)
    {
        if (std::find(names.begin(), names.end(), filter_input_column) == names.end())
            names.emplace_back(filter_input_column);
    }
```

Reply:
Your "binary search" doesn't seem to be faster than a linear search anyway. The complexity of a linear search is `O(N)`. You're using sort + binary search, the complexity is `O(N*log(N))`.

Reply:
Assuming there is one column to filter?